src/Frontend/Core.hs

```diff
@@ -247,13 +251,15 @@ redirectFormHandler getPage processor = getH :<|> postH
         v <- getForm fa (processor1 page)
         renderer page v fa
 
-    postH env = do
+    postH formData = do
         page <- getPage
         let fa = absoluteUriPath $ formAction page
+            env = getFormDataEnv formData
         (v, mpayload) <- postForm fa (processor1 page) (\_ -> return $ return . runIdentity . env)
-        case mpayload of
+        (case mpayload of
             Just payload -> processor2 page payload >>= redirect
-            Nothing      -> renderer page v fa
+            Nothing      -> renderer page v fa)
+            `finally` cleanupTempCsvFiles formData
```

Comment:
Am I right it's the `processor` argument in which the call to `popTempCsvFile` place?

It would be great if we could ensure non-laziness of the IO with `deepseq`, but I think we would need to hunt down any data resulting from `process` and call `deepseq` on it, or ensure strictness of all such data via types. The latter usually involves changing tons of code to only use strict types. I don't know the details of our data flow, but I think the type doesn't prevent the `process` argument from reading files and placing their content in state (persistent or not) of the system lazily without any relation to `payload`. Do we have guarantees the fields of our app state, whether the persistent part or some transient state, is strict? Additionally, are state updates in our monad strict enough (they usually are not)?